포커게임 결과 화면

TTOzzi opened the pull request:
SegmentedControl 스타일을 예시의 스타일처럼 바꿔보려 했는데 여러 예시코드를 따라 해봐도 적용이 안 되는 경우가 많았습니다. 뭔가 다른 곳에서 설정을 변경해줘서 안 바뀌는 것 같은데 너무 어렵습니다..
제가 봐도 제 코드가 깔끔하지 않은 것 같은데 함수들의 위치, 함수의 이름, 함수들을 어떤 기준으로 나눠야 할지가 가장 큰 고민거리였습니다. 제 나름대로 나눠주긴 했는데 잘 나눈 것 같지가 않습니다.

TTOzzi committed:
enum 선언 위치 수정

TTOzzi committed:
PokerGame 클래스 NumberOfPlayers enum 생성, 테스트코드 수정

게임 참여자 수를 enum 으로 만들어 선택지를 제한해줌
수정사항에 맞춰서 테스트코드 수정

TTOzzi committed:
Assets 에 카드 이미지 추가

TTOzzi committed:
Card 클래스 Shape rawValue 수정

이미지 파일명을 카드 객체와 맞추기위해 Shape 의 rawValue 알파벳으로 수정

TTOzzi committed:
카드 가로세로 비율 고정

constraint 로 가로 세로 비율을 1 : 1.27 로 고정해줌

TTOzzi committed:
Player 클래스에 hand의 정보를 얻기위한 handInfo 생성

hand 는 private으로 유지하고 hand 를 조회하기위해 만들어줌

TTOzzi committed:
화면 카드 출력 구현

pokerGame 의 게임타입, 참여자 수에 맞게 화면에 카드를 출력하도록 구현해줌
카드를 카드스택에 담고, 카드스택을 플레이어 스택에 담은 형태

TTOzzi committed:
플레이어, 딜러 표시 Label 추가

label 을 만들어 플레이어 순서에 맞게 텍스트를 넣은 뒤 playerStack 에 추가해줌
constraint 로 스택뷰의 spacing 에 영향을 받지 않고 카드스택 위에 붙어있도록 해줌

TTOzzi committed:
SegmentedControl 추가

gameType 을 표시해줄 gameTypeControl 과 참가자 수를 표시해줄 numberOfPlayersControl 추가

TTOzzi committed:
resetPlayerStack() 구현

게임타입이나 참가자 수가 바뀔 때 스택뷰를 리셋해주는 함수 구현

TTOzzi committed:
gameTypeControl, numberOfPlayersControl 액션 구현

스위치가 바뀔 때 마다 그에 맞춰서 화면이 바뀌도록 구현

TTOzzi committed:
코드 구조 및 함수, 변수 이름 변경

일정한 간격을 주기위해 플레이어 라벨과 플레이어 카드를 스택에 담아 게임테이블에 담아줌
코드들을 의미에 맞게 이름 변경

TTOzzi committed:
흔들어서 카드 섞기 구현

motionBegan 을 override 하여 shake 이벤트 발생시 loadGame() 을 호출하도록 구현

TTOzzi committed:
makeSegments 추가

SegmentedControl 을 생성할때 중복코드를 makeSegments 함수로 합쳐줌

TTOzzi committed:
ReadMe 업데이트

TTOzzi committed:
segmentedControl 에 border 추가

TTOzzi committed:
불필요 코드 제거

스택에 담아주니 자동으로 라벨과 카드를 붙여줘서 제거해줌

godrm commented:
우선 SegmentedControl 스타일은 너무 신경쓰지마세요. iOS 버전이 바뀔때 마다 조금씩 바뀝니다.

godrm commented on CardGameApp/CardGameApp/Card.swift:
```diff
@@ -9,21 +9,19 @@
-    }
 }
 
 extension Card: CustomStringConvertible, Equatable {
```
여기 Card == 비교에서는 그냥 그 값의 속성을 비교하도록 개선하세요.
문자열로 바꿔서 비교하는 것은 좋은 방법이 아닙니다.

godrm commented on CardGameApp/CardGameApp/Card.swift:
```diff
@@ -9,21 +9,19 @@
 
 class Card {
+    enum Shape: String, CaseIterable {
+        case heart = "h", spade = "s", clover = "c", diamond = "d"
```
유니코드로 표현했던 것은 왜 String으로 바꿨는지 궁금합니다.

TTOzzi replied:
카드 클래스 CustomStringConvertible 프로토콜의 description 을 카드 이미지명과 일치시켜주기위해 바꿔주었습니다.

godrm commented on CardGameApp/CardGameApp/Player.swift:
```diff
@@ -13,6 +13,9 @@ class Player {
     }
+    var handInfo: [Card] {
+        return hand
+    }
```
내부 속성을 선언할 때는 우선적으로 private var로 선언하세요. 
이렇게 값을 가져가는 getter 형식 메소드가 있다는 것은, 
enum, struct, class 값을 가져가서 비교하거나, 계산하거나, 탐색하는 코드가 객체 외부에 있다는 의미입니다.
그런 코드를 객체 내부에 구현하고 외부에서는 간단하게 호출하는 형태가 되도록 만드세요.

TTOzzi replied:
hand 안의 카드의 정보를 viewController 에서 imageView 로 만들기 위해 hand 를 반환하는 연산 프로퍼티를 만들어줬는데 말씀해주신대로 이 동작을 객체 내부에서 구현하려면 imageView 를 만들어 반환하도록 구현해야 할 텐데 이러면 모델의 역할이 불분명해지지 않나요..?

godrm replied:
아닙니다. Player가 ImageView까지 만들 필요없습니다. 
아래 힌트를 줬던 클로저를 응용해보세요.

TTOzzi replied:
아! 넵 감사합니다

godrm commented on CardGameApp/CardGameApp/PokerGame.swift:
```diff
@@ -9,21 +9,26 @@
         self.gameType = gameType
         self.dealer = Dealer()
-        for _ in 1...numberOfPlayers {
+        for _ in 1...numberOfPlayers.rawValue {
```
enum 타입도 rawValue에 접근하지 않고 처리할 수 있는 방법이 있습니다. 

클로저를 매개변수로 전달하면 내부 값을 가져오지 않아도 외부에서 원하는 로직을 전달할 수 있습니다.
```
enum NumberOfPlayers: Int {
    case two = 2
     case three = 3
     case four = 4
     
func foreach(_ transform : () -> ()) {
    for _ in 0..<self.rawValue {
        transform()
    }
}

}
```
호출할 때는 numberOfPlayers.foreach { self.players.append(Player()) } 형태가 됩니다.

내부의 속성, rawValue 값을 가져와서 처리하도록 하기 보다는 객체의 메소드를 호출하도록 만드는 게 좋습니다.

godrm commented on CardGameApp/CardGameApp/ViewController.swift:
```diff
@@ -25,20 +31,157 @@ class ViewController: UIViewController {
+        return cardStack
+    }
+    
+    func makeCardStack() -> UIStackView {
```
ViewController 내부에서만 사용하는 함수들은 private func 으로 지정하는 게 좋습니다.

godrm commented on CardGameApp/CardGameApp/ViewController.swift:
```diff
@@ -25,20 +31,157 @@ class ViewController: UIViewController {
+            loadGame()
+        case 2:
+            numberOfPlayers = .four
+            loadGame()
```
loadGame() 중복되는 데 switch-case 바깥에 한 줄로 넣어도 되지 않을까요?

TTOzzi replied:
엇 그렇네요. 수정하겠습니다.

godrm commented on CardGameApp/CardGameApp/ViewController.swift:
```diff
@@ -25,20 +31,157 @@ class ViewController: UIViewController {
+    @objc func numberOfPlayersChanged(segControl: UISegmentedControl) {
+        switch segControl.selectedSegmentIndex {
+        case 0:
+            numberOfPlayers = .two
```
이런 경우에 numberOfPlayers 를 만들기 위해서 switch-case 구문을 사용하는 것보다는
enum 타입의 생성자를 활용하는 게 좋습니다. 그래야 실수를 줄일 수 있고, 컴파일 타임에 확인할 수 있게 됩니다.

godrm commented on CardGameApp/CardGameApp/ViewController.swift:
```diff
@@ -25,20 +31,157 @@ class ViewController: UIViewController {
+    }
+    
+    override func motionBegan(_ motion: UIEvent.EventSubtype, with event: UIEvent?) {
+        loadGame()
```
터치 이벤트가 아니라 shake 이벤트입니다. 시뮬레이터에서 shake event를 찾아보세요.

godrm commented on CardGameApp/CardGameApp/ViewController.swift:
```diff
@@ -50,6 +50,10 @@ class ViewController: UIViewController {
+    override func motionBegan(_ motion: UIEvent.EventSubtype, with event: UIEvent?) {
+        loadGame()
+    }
+    
```
이벤트 처리 코드와 아래 뷰를 생성하는 코드를 배치를 시각적으로 구분해보세요.
`//MARK:-` 같은 표현도 가능합니다.

TTOzzi committed:
Card 클래스 Equatable 프로토콜 수정

카드 내부 속성을 모두 비교해주도록 수정

TTOzzi committed:
PokerGame 클래스 GameType, NumberOfPlayers 수정

rawValue 에 접근하지않고 enum 내부에서 메소드를 호출하도록 변경

TTOzzi committed:
Player 클래스 handInfo 삭제

객체 내부에서 값을 끄집어내는게 아닌 메소드를 외부에서 받아와서 객체 내부에서 처리해주도록 변경

TTOzzi committed:
PokerGame 클래스 players pirvate 설정

players 를 private 으로 설정해주고 forEachPlayer 로 플레이어가 필요한 로직을 외부에서 가져와서 수행하도록 해줌

TTOzzi committed:
PokerGame 클래스 GameType, NumberOfPlayers enum 생성자 생성

TTOzzi committed:
SegmentedControl 타켓함수 수정

enum 생성자를 활용해 selectedSegmentIndex 로 case 를 찾아 변경해주도록 수정

TTOzzi committed:
라벨 제작에서 딜러인지 확인하는 로직 수정

index 와 player 의 count 를 비교해서 딜러를 확인했었는데 type 이 딜러인지 확인하는 방식으로 변경
PokerGame 클래스의 playersCount 가 필요없어져 없애줌

godrm commented:
모두 개선한 걸까요? 기다려야 할까요? 댓글을 남겨야 제가 판단할 수 있습니다.

TTOzzi commented:
아직 수정중이에요! 잠시만 기다려주세요!!!

godrm commented on CardGameApp/CardGameApp/Player.swift:
```diff
@@ -13,10 +13,10 @@ class Player {
     
+    func forEach(_ tramsform: (Card) -> ()) {
+        hand.forEach(tramsform)
+    }
```
`transform` 일 것 같은데 `tramsform` 이네요 :) 
이런 방식일 때는 매개변수로 카드 객체를 그대로 넘기는 게 좋은지는 또 다른 고민해볼 포인트입니다. 
지금은 Card가 공통적인 데이터 구조라 괜찮아 보입니다.

godrm commented on CardGameApp/CardGameApp/PokerGame.swift:
```diff
@@ -28,12 +48,16 @@ class PokerGame {
+    }
+    
     func distributeCards() {
         players.forEach {
```
이제 이쯤에서는 players를 탐색하는 코드가 여기 게임에 있을 필요가 없다는 것을 알아채면 좋겠습니다.
players 배열을 직접 참조하기 보다는 Players 라는 객체가 별도로 있는 것은 어떨까요?
그러면 여기서 탐색하고 처리하는 코드는 Players 내부에 있고, 마찬가지 방식으로 여기서 클로저로 넘기도록 할 수 있습니다. 
다음 단계를 하면서 같이 고민해보세요.

godrm commented:
마무리하고 댓글로 남겨주세요. 
그리고 push는 모두 수정 완료하고 하는 게 좋습니다 :)

TTOzzi commented:
넵 알겠습니다!

TTOzzi committed:
Pragma Mark 로 뷰 생성부분과 이벤트 처리부분 분리, 오타수정

TTOzzi commented:
마무리했습니다!
Player 객체 분리는 다음 단계에서 생각해보겠습니다. 감사합니다.
